### PR TITLE
Introduce utils for _really_ stashing the thread context

### PR DESCRIPTION
**Pull request by DaveCTurner**
`ThreadContext#stashContext` does not yield a completely fresh context:
it preserves headers related to tracing the original request. That may
be appropriate in many situations, but sometimes we really do want to
detach processing entirely from the original task. This commit
introduces new utilities to do that.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 1]
```diff
@@ -109,7 +109,8 @@ private void messageReceived(TcpChannel channel, InboundMessage message, long st
 
         TransportResponseHandler<?> responseHandler = null;
         ThreadContext threadContext = threadPool.getThreadContext();
-        try (ThreadContext.StoredContext existing = threadContext.stashContext()) {
+        assert threadContext.isDefaultContext();
+        try (var ignored = threadContext.newStoredContext()) {
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:113]: Just checking: why is this changed?
I assume we don't care about preserving headers here, as we set them explicitly just at the next step?

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java [hunk 2]
```diff
@@ -237,7 +236,7 @@ private void internalSend(
         final long messageSize = reference.length();
         TransportLogger.logOutboundMessage(channel, reference);
         // stash thread context so that channel event loop is not polluted by thread context
-        try (ThreadContext.StoredContext existing = threadPool.getThreadContext().stashContext()) {
+        try (var ignored = threadPool.getThreadContext().newEmptyContext()) {
```
Comment (ldematte) [inline server/src/main/java/org/elasticsearch/transport/OutboundHandler.java:239]: Same question: why this changed?

**Review by ldematte — COMMENTED** [review]
A couple of Qs on changes that are not straightforward replacements (I think)

**Review by rjernst — COMMENTED** [review]
Aside from Lorenzo's questions, this looks great to me. The new methods are much clearer and more succinct.

**Review by rjernst — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/InboundHandler.java:113]: Yes right, no need to muck around creating the stashed context when we're overwriting its contents straight away. But anyway it's a bug not to be in default context here so there's no difference.

File: server/src/main/java/org/elasticsearch/transport/OutboundHandler.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/transport/OutboundHandler.java:239]: Shouldn't really matter either way but we generally expect the thread context to be totally empty on the Netty event loop. This was one case where stashing used to do the right thing but today might leak tracing-related headers.
